[WebGPU] Flesh out PresentationContext's API

litherum opened the pull request:
#### 7d48787400be8e0bc45e3bf0c0817d510341347f
<pre>
[WebGPU] Flesh out PresentationContext&apos;s API
<a href="https://bugs.webkit.org/show_bug.cgi?id=250990">https://bugs.webkit.org/show_bug.cgi?id=250990</a>
rdar://104537242

Reviewed by Tadeu Zagallo.

This patch does 3 things:
1. Migrates getCurrentTextureView() to getCurrentTexture(). GPUCanvasContext uses
       getCurrentTexture(), but WebGPU.h exposes getCurrentTextureView(). GPUCanvasContext
       represents the source of truth; WebGPU.h is outdated, so we should use
       getCurrentTexture(). webgpu-native/webgpu-headers#89 is
       about updating WebGPU.h.
2. Implements getCurrentTexture(). These are a little tricky because they&apos;re supposed to be
       +0 getters. However, that&apos;s only relevant for calling directly into WebGPU.framework;
       everywhere else in WebKit we use owning Refs, so in WebKit, a +0 object is really just
       an object whose creator has a ref to it. The tricky parts about dealing with
       WebGPU.framework will be fixed in <a href="https://bugs.webkit.org/show_bug.cgi?id=250958">https://bugs.webkit.org/show_bug.cgi?id=250958</a>; this
       patch hooks just those parts up intentionally incorrectly and adds a FIXME with a link
       to that bug. These functions aren&apos;t called yet (that&apos;s for
       <a href="https://bugs.webkit.org/show_bug.cgi?id=250995)">https://bugs.webkit.org/show_bug.cgi?id=250995)</a> so it&apos;s okay that this part doesn&apos;t
       actually work yet. WebGPU.framework continues to implement getCurrentTextureView()
       just for completeness.
3. Adds present(), and hooks it up to wgpuSwapChainPresent(). This part is super
       straightforward.

* Source/WebCore/Modules/WebGPU/GPUPresentationContext.cpp:
(WebCore::GPUPresentationContext::getCurrentTexture):
(WebCore::GPUPresentationContext::present):
* Source/WebCore/Modules/WebGPU/GPUPresentationContext.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUPresentationContextImpl.cpp:
(PAL::WebGPU::PresentationContextImpl::configure):
(PAL::WebGPU::PresentationContextImpl::getCurrentTexture):
(PAL::WebGPU::PresentationContextImpl::present):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUPresentationContextImpl.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUPresentationContext.h:
* Source/WebCore/html/canvas/GPUCanvasContextCocoa.cpp:
(WebCore::GPUCanvasContextCocoa::getCurrentTexture):
* Source/WebGPU/WebGPU/PresentationContext.h:
* Source/WebGPU/WebGPU/PresentationContext.mm:
(WebGPU::PresentationContext::getCurrentTexture):
(wgpuSwapChainGetCurrentTexture):
* Source/WebGPU/WebGPU/PresentationContextCoreAnimation.h:
(WebGPU::PresentationContextCoreAnimation::Configuration::Configuration):
* Source/WebGPU/WebGPU/PresentationContextCoreAnimation.mm:
(WebGPU::PresentationContextCoreAnimation::configure):
(WebGPU::PresentationContextCoreAnimation::Configuration::generateCurrentFrameState):
(WebGPU::PresentationContextCoreAnimation::getCurrentTexture):
* Source/WebGPU/WebGPU/PresentationContextIOSurface.h:
* Source/WebGPU/WebGPU/PresentationContextIOSurface.mm:
(WebGPU::PresentationContextIOSurface::getCurrentTexture):
* Source/WebGPU/WebGPU/WebGPUExt.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemotePresentationContext.cpp:
(WebKit::RemotePresentationContext::getCurrentTexture):
(WebKit::RemotePresentationContext::present):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemotePresentationContext.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemotePresentationContext.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.cpp:
(WebKit::WebGPU::RemoteDeviceProxy::createTexture):
(WebKit::WebGPU::RemoteDeviceProxy::createSurfaceTexture):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemotePresentationContextProxy.cpp:
(WebKit::WebGPU::RemotePresentationContextProxy::getCurrentTexture):
(WebKit::WebGPU::RemotePresentationContextProxy::present):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemotePresentationContextProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteTextureProxy.cpp:
(WebKit::WebGPU::RemoteTextureProxy::RemoteTextureProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteTextureProxy.h:

Canonical link: <a href="https://commits.webkit.org/259585@main">https://commits.webkit.org/259585@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8e2506e9d2fa548082be2005316935f43fe60a8a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105327 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14403 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38208 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114584 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/174768 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109230 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/15594 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5332 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97636 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/114490 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111085 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/15594 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/95032 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/39533 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/76/builds/15594 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26666 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/81203 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/7728 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/28026 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7837 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/4613 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13872 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47570 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6606 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9624 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteTextureViewProxy.cpp:
```diff
@@ -33,16 +33,14 @@
 {
 }
 
-RemoteTextureViewProxy::~RemoteTextureViewProxy()
```
this change isn't really necessary in this PR, and will be un-done by https://github.com/WebKit/WebKit/pull/8856

litherum replied:
🆗

mwyrzykowski commented on Source/WebCore/Modules/WebGPU/GPUPresentationContext.cpp:
```diff
@@ -41,10 +41,39 @@ void GPUPresentationContext::unconfigure()
+
+void GPUPresentationContext::ensureCurrentTextureAndView()
+{
+    ASSERT(static_cast<bool>(m_currentTexture) == static_cast<bool>(m_currentTextureView));
```
could also be written as: `ASSERT(!!m_currentTexture == !!m_currentTextureView);` 

!! is pretty standard for conversion to bool throughout webkit and much shorter.

litherum replied:
Ugh I hate `!!` :( It sacrifices clarity for brevity

litherum replied:
I'm going to intentionally not abide by this review comment 🤔

litherum replied:
Oh, conveniently, this code goes away when I update for https://github.com/WebKit/WebKit/pull/9263#issuecomment-1407257039

mwyrzykowski replied:
👍 

Also totally fine to ignore the !! convention if it comes up in the future, just a suggestion :)

mwyrzykowski commented on Source/WebCore/Modules/WebGPU/GPUPresentationContext.cpp:
```diff
@@ -41,10 +41,39 @@ void GPUPresentationContext::unconfigure()
+        return;
+
+    m_currentTexture = GPUTexture::create(m_backing->getCurrentTexture()).ptr();
+    m_currentTextureView = GPUTextureView::create(m_backing->getCurrentTextureView()).ptr();
```
I don't see this used anywhere by the web API. Won't this result in creating an extra texture view every frame?

mwyrzykowski reviewed (changes requested):
I don't think we should merge this until we clear up the disconnect with the Web API.

Will getCurrentTextureView() exist in the future? It doesn't exist in the Web API. From the discussions in webgpu-native it sounds like WebGPU.h has fallen behind with the web API.

Note the PR here removes getCurrentTextureView in place of getCurrentTexture https://github.com/webgpu-native/webgpu-headers/pull/164/files which it sounds like will not be addressed until post-v1.

mwyrzykowski reviewed (approved):
Looks good, but commit message still references changes to getCurrentTextureView

litherum committed:
[WebGPU] Flesh out PresentationContext's API
https://bugs.webkit.org/show_bug.cgi?id=250990
rdar://104537242

Reviewed by Tadeu Zagallo.

This patch does 3 things:
1. Migrates getCurrentTextureView() to getCurrentTexture(). GPUCanvasContext uses
       getCurrentTexture(), but WebGPU.h exposes getCurrentTextureView(). GPUCanvasContext
       represents the source of truth; WebGPU.h is outdated, so we should use
       getCurrentTexture(). webgpu-native/webgpu-headers#89 is
       about updating WebGPU.h.
2. Implements getCurrentTexture(). These are a little tricky because they're supposed to be
       +0 getters. However, that's only relevant for calling directly into WebGPU.framework;
       everywhere else in WebKit we use owning Refs, so in WebKit, a +0 object is really just
       an object whose creator has a ref to it. The tricky parts about dealing with
       WebGPU.framework will be fixed in https://bugs.webkit.org/show_bug.cgi?id=250958; this
       patch hooks just those parts up intentionally incorrectly and adds a FIXME with a link
       to that bug. These functions aren't called yet (that's for
       https://bugs.webkit.org/show_bug.cgi?id=250995) so it's okay that this part doesn't
       actually work yet. WebGPU.framework continues to implement getCurrentTextureView()
       just for completeness.
3. Adds present(), and hooks it up to wgpuSwapChainPresent(). This part is super
       straightforward.

* Source/WebCore/Modules/WebGPU/GPUPresentationContext.cpp:
(WebCore::GPUPresentationContext::getCurrentTexture):
(WebCore::GPUPresentationContext::present):
* Source/WebCore/Modules/WebGPU/GPUPresentationContext.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUPresentationContextImpl.cpp:
(PAL::WebGPU::PresentationContextImpl::configure):
(PAL::WebGPU::PresentationContextImpl::getCurrentTexture):
(PAL::WebGPU::PresentationContextImpl::present):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUPresentationContextImpl.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUPresentationContext.h:
* Source/WebCore/html/canvas/GPUCanvasContextCocoa.cpp:
(WebCore::GPUCanvasContextCocoa::getCurrentTexture):
* Source/WebGPU/WebGPU/PresentationContext.h:
* Source/WebGPU/WebGPU/PresentationContext.mm:
(WebGPU::PresentationContext::getCurrentTexture):
(wgpuSwapChainGetCurrentTexture):
* Source/WebGPU/WebGPU/PresentationContextCoreAnimation.h:
(WebGPU::PresentationContextCoreAnimation::Configuration::Configuration):
* Source/WebGPU/WebGPU/PresentationContextCoreAnimation.mm:
(WebGPU::PresentationContextCoreAnimation::configure):
(WebGPU::PresentationContextCoreAnimation::Configuration::generateCurrentFrameState):
(WebGPU::PresentationContextCoreAnimation::getCurrentTexture):
* Source/WebGPU/WebGPU/PresentationContextIOSurface.h:
* Source/WebGPU/WebGPU/PresentationContextIOSurface.mm:
(WebGPU::PresentationContextIOSurface::getCurrentTexture):
* Source/WebGPU/WebGPU/WebGPUExt.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemotePresentationContext.cpp:
(WebKit::RemotePresentationContext::getCurrentTexture):
(WebKit::RemotePresentationContext::present):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemotePresentationContext.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemotePresentationContext.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.cpp:
(WebKit::WebGPU::RemoteDeviceProxy::createTexture):
(WebKit::WebGPU::RemoteDeviceProxy::createSurfaceTexture):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemotePresentationContextProxy.cpp:
(WebKit::WebGPU::RemotePresentationContextProxy::getCurrentTexture):
(WebKit::WebGPU::RemotePresentationContextProxy::present):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemotePresentationContextProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteTextureProxy.cpp:
(WebKit::WebGPU::RemoteTextureProxy::RemoteTextureProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteTextureProxy.h:

Canonical link: https://commits.webkit.org/259585@main
